TypeError when trying to use SSL on PyPy3

In a `pypy:3` Docker container using PyPy3 7.3.3, pytest 6.2.2, and smtpdfix 0.2.9, the following code fails with a `TypeError`:

```python
from email.message import EmailMessage
import smtplib
import pytest

@pytest.fixture
def mock_smtpd_use_ssl(monkeypatch):
    monkeypatch.setenv("SMTPD_USE_SSL", "True")

def test_smtpd_ssl(mock_smtpd_use_ssl, smtpd):
    client = smtplib.SMTP_SSL(smtpd.hostname, smtpd.port)
    msg = EmailMessage()
    msg["Subject"] = "Hello, World!"
    msg["To"] = "world@earth.space"
    msg["From"] = "me@here.qq"
    msg.set_content("Hi, world!  How you doin'?\n")
    client.send_message(msg)
    client.quit()
    assert len(smtpd.messages) == 1
```

The output from pytest is as follows:

```
============================= test session starts ==============================
platform linux -- Python 3.6.12[pypy-7.3.3-final], pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /docked
plugins: smtpdfix-0.2.9
collected 1 item

test_ssl2.py E                                                           [100%]

==================================== ERRORS ====================================
_______________________ ERROR at setup of test_smtpd_ssl _______________________

tmp_path_factory = TempPathFactory(_given_basetemp=None, _trace=<pluggy._tracing.TagTracerSub object at 0x00007f0eab940950>, _basetemp=PosixPath('/tmp/pytest-of-root/pytest-2'))

    @pytest.fixture
    def smtpd(tmp_path_factory):
        """A small SMTP server for use when testing applications that send email
        messages. To access the messages call `smtpd.messages` which returns a copy
        of the list of messages sent to the server.
    
        Example:
            def test_mail(smtpd):
                from smtplib import SMTP
                with SMTP(smtpd.hostname, smtpd.port) as client:
                    code, resp = client.noop()
                    assert code == 250
        """
        config = Config()
    
        if os.getenv("SMTPD_SSL_CERTS_PATH") is None:
            path = tmp_path_factory.mktemp("certs")
            generate_certs(path)
            os.environ["SMTPD_SSL_CERTS_PATH"] = str(path.resolve())
    
        # fixture = AuthController(
        #     hostname=config.SMTPD_HOST,
        #     port=int(config.SMTPD_PORT),
        #     config=config,
        #     authenticator=_Authenticator(config)
        # )
        # fixture.start()
        # yield fixture
        # fixture.stop()
    
>       with SMTPDFix(config.SMTPD_HOST, config.SMTPD_PORT) as fixture:

/opt/pypy/site-packages/smtpdfix/fixture.py:105: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/pypy/site-packages/smtpdfix/fixture.py:64: in __enter__
    authenticator=_Authenticator(self.config)
/opt/pypy/site-packages/smtpdfix/controller.py:48: in __init__
    ssl_context=context_or_none())
/opt/pypy/site-packages/smtpdfix/controller.py:40: in context_or_none
    return ssl_context or self._get_ssl_context()
/opt/pypy/site-packages/smtpdfix/controller.py:81: in _get_ssl_context
    context.load_cert_chain(cert_path, key_path)
/opt/pypy/lib_pypy/_cffi_ssl/_stdssl/__init__.py:1113: in load_cert_chain
    certfilebuf = _str_to_ffi_buffer(certfile)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

view = PosixPath('/tmp/pytest-of-root/pytest-2/certs0/cert.pem')

    def _str_to_ffi_buffer(view):
        if isinstance(view, str):
            return ffi.from_buffer(view.encode())
        elif isinstance(view, memoryview):
            # NOTE pypy limitation StringBuffer does not allow
            # to get a raw address to the string!
            view = view.tobytes()
        # dont call call ffi.from_buffer(bytes(view)), arguments
        # like ints/bools should result in a TypeError
>       return ffi.from_buffer(view)
E       TypeError: a bytes-like object is required, not PosixPath

/opt/pypy/lib_pypy/_cffi_ssl/_stdssl/utility.py:25: TypeError
=========================== short test summary info ============================
ERROR test_ssl2.py::test_smtpd_ssl - TypeError: a bytes-like object is requir...
=============================== 1 error in 0.51s ===============================
```

The fixture works on PyPy3 when SSL is not used.  When StartTLS is used on PyPy3, pytest just hangs indefinitely; I suspect that the `TypeError` is killing the SMTPD job and nothing's cleaning it up.

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -13,6 +13,7 @@ As of version 0.2.10 smtpdfix uses the standard aiosmtpd VRFY handler that will 
 - Better compatibility with AUTH in aiosmtpd by using `AuthResult` as return from AUTH mechanism handlers and the `Authenticator` for processing.
 - Uses default aiosmtpd VRFY handling and responses.
 - Adds unofficial support for PyPy3 in response to. [Issue #51](https://github.com/bebleo/smtpdfix/issues/51)
+- Adds exception handling to prevent cases where unhandled exceptions cause the fixture to hang. Now returns a 421 error and closes the connection. [Issue #51](https://github.com/bebleo/smtpdfix/issues/51)
 
 ### Contributions
 

--- a/smtpdfix/controller.py
+++ b/smtpdfix/controller.py
@@ -1,3 +1,4 @@
+import asyncio
 import errno
 import logging
 import ssl
@@ -30,6 +31,8 @@ class AuthController(Controller):
         _handler = AuthMessage(messages=self._messages)
         _hostname = hostname or self.config.SMTPD_HOST
         _port = int(port or self.config.SMTPD_PORT or 8025)
+        _loop = loop or asyncio.new_event_loop()
+        _loop.set_exception_handler(self.handle_exception)
 
         def context_or_none():
             # Determines whether to return a sslContext or None to avoid a
@@ -45,7 +48,7 @@ class AuthController(Controller):
         super().__init__(handler=_handler,
                          hostname=_hostname,
                          port=_port,
-                         loop=loop,
+                         loop=_loop,
                          ready_timeout=ready_timeout,
                          ssl_context=context_or_none(),
                          authenticator=self._authenticator)
@@ -82,12 +85,20 @@ class AuthController(Controller):
                                     file_)
 
         context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
-        
+
         # Becuase PYPY3 doesn't support Path objects when loading the
         # certificate and keys we cast them to a string.
         context.load_cert_chain(str(cert_path), keyfile=str(key_path))
 
         return context
+
+    def handle_exception(self, loop, context):
+        loop.default_exception_handler(context)
+
+        status = "421 Service not available. Closing connection."
+        asyncio.ensure_future(self.smtpd.push(status))
+        self.smtpd.transport.close()
+        self.server.close()
 
     @property
     def messages(self):

--- a/tests/test_controller.py
+++ b/tests/test_controller.py
@@ -101,3 +101,20 @@ async def test_config_file(request, msg):
 
     os.environ.clear()
     os.environ.update(_original_env)
+
+
+async def test_exception_handler(request, msg):
+    def raise_error():
+        raise Exception("Deliberately raised error.")
+
+    server = AuthController()
+    request.addfinalizer(server.stop)
+    server.start()
+
+    with pytest.raises(Exception):
+        with SMTP(server.hostname, server.port) as client:
+            client.ehlo()
+            server.loop.call_soon_threadsafe(raise_error)
+            client.send_message(msg)
+
+        assert len(server.messages) == 1

